File: 238-product-of-array-except-self/product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> result(nums.size(), 1);
        int buf = 1;
        for (auto i = 1; i < nums.size(); i++) 
            result[i] = result[i - 1] * nums[i - 1];
         
        for (int i = nums.size() - 1; i >= 0; i--) {
            result[i] *= buf;
            buf *= nums[i];
        }
        return result;
    }
};

File: 238-product-of-array-except-self/product-of-array-except-self.go
func productExceptSelf(nums []int) []int {
    n := len(nums)
    res := make([]int, n)
    res[0] = 1

    for i := 1; i < n; i++ {
        res[i] = res[i - 1] * nums[i - 1]
    }

    x := 1
    for i := n - 1; i >= 0; i-- {
        res[i] = res[i] * x
        x *= nums[i]
    }

    return res
}

